[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63828
* #63627
* #63501
* __->__ #62861
* #62698

Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

jerryzh168 committed:
[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu

Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

vkuzo commented on torch/quantization/fx/lower_to_native_backend.py:
```diff
@@ -0,0 +1,8 @@
+import torch
+
+def lower_to_native_backend(model: torch.fx.GraphModule):
```
should the type be `nn.Module`, and we trace again?

should we create `lower_to_fbgemm` and `lower_to_qnnpack`, even if the code is the same for now?  They may diverge in the future, and this way the API is more future proof.

jerryzh168 replied:
Sure, we can trace again.

Yeah I can split the function into 2.

jerryzh168 replied:
actually if we trace again we might need to carry over configurations from quantization, which is a little bit cumbersome, can we use this as an internal function and just accept torch.fx.GraphModule instead?

raghuramank100 replied:
I think it is ok to have this be an internal function as users interface with this only via convert_fx.

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

raghuramank100 commented on torch/quantization/fx/convert.py:
```diff
@@ -45,6 +45,8 @@
     activation_dtype,
 )
 
+from .lower_to_fbgemm import lower_to_fbgemm
```
The convert function here (L145) is almost 400 lines long. Can you refactor it in a later PR? With multiple functions defined inside, it is hard to understand the flow here.

jerryzh168 replied:
yes sure, will do

jerryzh168 replied:
I think we also might be able to simplify a bit after we change the flow to produce reference pattern by default

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -1496,7 +1496,9 @@ def convert(self,
                 is_reference: bool = False,
                 convert_custom_config_dict: Dict[str, Any] = None) -> Node:
-        if is_reference:
+        # always produce reference pattern for relu
```
Why should we do this? If we lower to fbgemm or qnnpack, isnt this incorrect?

jerryzh168 replied:
oh, this is to change convert_fx always to produce reference pattern, and we'll fuse the pattern in lower_to_fbgemm step

vkuzo replied:
would this break anything for customers who are not using the new `lower_to_fbgemm` function yet?

jerryzh168 replied:
no, this is internal implementation change, users are not explicitly calling lower_to_fbgemm currently. (they just need to call convert_fx)

vkuzo replied:
I'm probably confused, let's say a customer has a relu in their model, is_reference is False, and they are not using any lowering functions.  Would this PR change the behavior of calling convert on their model?  From the code it seems like yes, but I might be missing something.

jerryzh168 replied:
no, currently user do not need to call lowering function separately, it's called in convert_fx when is_reference=False.

vkuzo replied:
makes sense, thanks!

raghuramank100 commented on torch/quantization/fx/lower_to_native_backend.py:
```diff
@@ -0,0 +1,8 @@
+    to the native backend in PyTorch (fbgemm/qnnpack), both backends shares the same
+    operator signature so they can be lowered with the same function
+    """
+    pass
```
Currently these functions are no-ops. So how do we go from reference to actual backend?

jerryzh168 replied:
sorry, the changes are lost somehow, I just updated the PR again

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30165828).

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 commented on torch/quantization/fx/convert.py:
```diff
@@ -535,4 +537,5 @@ def load_arg_remove(a: Argument) -> Argument:
     model = QuantizedGraphModule(model, act_post_process_removed_graph, preserved_attributes)
     if not is_reference:
         model = fold_weight(model, node_name_to_scope)
+        model = lower_to_fbgemm(model)
```
we call the lowering function here in convert_fx, so no change of api for users with this PR

vkuzo replied:
makes sense, thanks.  Should we add a TODO here to lower to the correct choice from (fbgemm, qnnpack), and also verify that all qconfigs are valid for the selected backend from (fbgemm, qnnpack)?

vkuzo commented on torch/quantization/fx/quantized_fusion_patterns_and_replacements.py:
```diff
@@ -0,0 +1,31 @@
+import torch
+
+def relu_inplace_pattern(x, scale, zero_point):
```
would be great to have I/O typehints for these

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][graphmode][fx] Add a separate lower_to_native_backend function for relu"


Summary:
This PR adds a lower_to_native_backend function to lower a quantized reference model
to a model that uses fbgemm/qnnpack ops. We'll gradually add support and remove
the fbgemm/qnnpack specific handling in quantization_patterns.py

Test Plan:
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30165828](https://our.internmc.facebook.com/intern/diff/D30165828)

[ghstack-poisoned]
